### PR TITLE
8239589: JavaFX UI will not repaint after reconnecting via Remote Desktop

### PR DESCRIPTION
**Pull request by kevinrushforth**
This is a fix for a long-standing bug where the D3D pipeline will stop rendering when a Windows remote desktop session is disconnected and then reconnected.

A preliminary Draft PR #315 by @Schmidor was a good first step in solving this. I took that and continued the work in my Draft PR #403. It is now ready for formal review in this new PR. You can see PR #403 for details on the history of the changes.

## Evaluation

The root cause of this bug is that the D3D pipeline did not handle a return code of `D3DERR_DEVICEREMOVED` from `TestCooperativeLevel`. When that error occurs, an application needs to destroy and recreate the Direct3D device.

The solution is to implement a new `D3DPipeline::reinitialize` method that will destroy the native D3D device and dispose the existing ResourceFactory objects and their associated BaseContext objects upon receiving `D3DERR_DEVICEREMOVED`. Note that the `D3DPipeline` Java object singleton is not recreated (it remains a singleton). In support of this, I implemented proper disposal logic in `BaseResourceFactory` and `BaseContext` to clean everything up and also to avoid memory leaks.

Additionally, there were several places that assumed that some textures (and mesh vertices) could be made permanent and never need to handle the case of a lost device. These all had to be fixed to allow for the possibility of a lost device and associated resource factory. They included:

* UploadingPainter and PresentingPainter need to set the resource factory to null when not ready, so it will get the (possibly new) factory the next time it tries.
* The gradient texture cache in `PaintHelper` has to be cleared and recreated when the surface is lost
* The 3D triangle mesh and Phong material classes need to be disposed when the resource factory is disposed.
* WebView often renders to a texture image at a time other than from the main rendering job, so needs to directly handle the case of a resource factory that is lost.
* Decora PPSRenderer assumed that the resource factory never went away; it also accessed it on the wrong thread. Both problems were addressed by deferring the initialization of the resource factory and handling the case where the device is disposed.
* Snapshot needs to allow for the platform image to be null if the device has been disposed.

## Notes to Reviewers

I created this PR from a branch that contains the original 4 commits by @Schmidor (rebased on top of the current `master`) and then a single commit on top of that to complete it. This allows anyone who is interested to easily see the diffs between this PR and Oliver's original Draft PR. Most reviewers can just go to the list of "Files" and see the aggregate diffs.

During the course of my testing I discovered three outstanding problems, which will be handled by filing follow-up issues. Once I file them, I'll add a comment to this PR with the bug IDs.

1. Media: a media stream playing at the time of a reconnect doesn't continue playing. Reloading the media works fine. This is not directly related to this bug, since it also happens with the software pipeline.
2. Canvas: doesn't preserve the contents after a device reconnect (noticed while running Zoomy, where the BG color is wrong after device reinitialization). This might point to a need to let the app know they have to repaint, since there is no possible way to preserve the contents of the texture when the device is lost.
3. WebView: there is a possible memory leak when device isn't ready after first reset, due to a `WCRenderQueueImpl::gc` instance being held in a JNIGlobal. This looks like a preexisting condition that could happen with a page (re)load today. It happens rarely.

This is a complicated enough change that I'd like three reviewers. The bulk of the changes are Windows-specific, but there are changes in common code so at least a sanity check needs to be done on all platforms using both the HW and SW pipelines. The case of a disposed device can currently only happen on Windows with the D3D pipeline.

/reviewers 3
/contributor add Oliver Schmidtmer <O.Schmidtmer@elo.com>
/contributor add kcr

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8239589](https://bugs.openjdk.java.net/browse/JDK-8239589): JavaFX UI will not repaint after reconnecting via Remote Desktop

### Reviewers
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)
 * [Arun Joseph](https://openjdk.java.net/census#ajoseph) (@arun-Joseph - Committer)

### Contributors
 * Oliver Schmidtmer `<O.Schmidtmer@elo.com>`
 * Kevin Rushforth `<kcr@openjdk.org>`

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jfx pull/430/head:pull/430`
`$ git checkout pull/430`

To update a local copy of the PR:
`$ git checkout pull/430`
`$ git pull https://git.openjdk.java.net/jfx pull/430/head`

### CODE REVIEW

**Comment by kevinrushforth** [comment]
NOTE: the Windows GitHub actions build is failing due to [JDK-8259639](https://bugs.openjdk.java.net/browse/JDK-8259639). I'll merge that fix in from master once it is integrated.

**Comment by kevinrushforth** [comment]
@Schmidor if you are able to review or test this, it would be appreciated.

**Review by kevinrushforth — COMMENTED** [review]
I did a pass over the review and left a few inline comments pointing out minor changes that I plan to make (in print statements and comments).

**Comment by Schmidor** [comment]
I can confirm the reinitialization after RDP reconnect works for me, standalone as JFX application and embedded in Swing.
Even either the multiple reinit retries of earlier fix versions aren't necessary anymore or it is not showing in debug output.

**Comment by kevinrushforth** [comment]
Thanks for confirming. I got rid of all debugging code, so it wouldn't show in the output. If you want to see whether or not multiple re-inits are happening, you can run with the following env variable set:

```
    export NWT_TRACE_LEVEL=2
```

If you are getting a failure on first time init, it will say something like `Zero adapters found`.

File: modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/RTImage.java
Comment (arapte) [inline modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/RTImage.java:52]: may be rename the variable to `resourceFactoryRef`

**Review by arapte — COMMENTED** [review]
The fix and Sanity testing looks all good.
I shall need some more time to do a detailed review.
Added minor comments, please address whenever you update next commit.

File: modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/RTImage.java
Comment (kevinrushforth) [reply modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/RTImage.java]: I renamed it here, and in `WCPageBackBufferImpl`, to `registeredWithFactory` which is the same name as is used in `PrismMediaFrameHandler`.

File: modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/WCPageBackBufferImpl.java
Comment (arun-Joseph) [inline modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/WCPageBackBufferImpl.java:126]: Minor: factoryReleased -> factoryReleased()

**Comment by kevinrushforth** [comment]
The new unit test fails on all platforms, as you can see from the GitHub actions log:

```
2021-03-18T12:45:36.7904530Z test.javafx.scene.control.ControlAcceleratorSupportTest > sanityTestListenerMapShouldBeEmpty FAILED
2021-03-18T12:45:36.8007000Z     java.lang.AssertionError: expected:<0> but was:<22>
2021-03-18T12:45:36.8099570Z         at org.junit.Assert.fail(Assert.java:91)
2021-03-18T12:45:36.8201200Z         at org.junit.Assert.failNotEquals(Assert.java:645)
2021-03-18T12:45:36.8264450Z         at org.junit.Assert.assertEquals(Assert.java:126)
2021-03-18T12:45:36.8366120Z         at org.junit.Assert.assertEquals(Assert.java:470)
2021-03-18T12:45:36.8455230Z         at org.junit.Assert.assertEquals(Assert.java:454)
2021-03-18T12:45:36.8559230Z         at test.javafx.scene.control.ControlAcceleratorSupportTest.sanityTestListenerMapShouldBeEmpty(ControlAcceleratorSupportTest.java:43)
2021-03-18T12:45:36.8886110Z 
2021-03-18T12:45:36.8989430Z test.javafx.scene.control.ControlAcceleratorSupportTest > testNumberOfListenersByRemovingAndAddingMenuItems FAILED
2021-03-18T12:45:36.9092540Z     java.lang.AssertionError: expected:<4> but was:<26>
2021-03-18T12:45:36.9194300Z         at org.junit.Assert.fail(Assert.java:91)
2021-03-18T12:45:36.9296390Z         at org.junit.Assert.failNotEquals(Assert.java:645)
2021-03-18T12:45:36.9393550Z         at org.junit.Assert.assertEquals(Assert.java:126)
2021-03-18T12:45:36.9495860Z         at org.junit.Assert.assertEquals(Assert.java:470)
2021-03-18T12:45:36.9596430Z         at org.junit.Assert.assertEquals(Assert.java:454)
2021-03-18T12:45:36.9700420Z         at test.javafx.scene.control.ControlAcceleratorSupportTest.testNumberOfListenersByRemovingAndAddingMenuItems(ControlAcceleratorSupportTest.java:65)
```

**Review by arapte — APPROVED** [review]
Fix looks good to me. Sanity testing with all three d3d, es2, sw pipelines behave as expected. Apps work as expected in case of remote desktop, sleep, screen lock.

**Review by arun-Joseph — APPROVED** [review]
Fix looks good. Tested WebView by reconnecting via remote desktop.

**Comment by kevinrushforth** [comment]
@Schmidor did you have any further comments or feedback?

Btw, I realized that you don't have a role in the openjfx project, so you can't formally review it. As such, I'll change the requirements back to 2 reviewers. I'll hold off integrating it until tomorrow in case you have additional feedback.

/reviewers 2

**Comment by Schmidor** [comment]
@kevinrushforth No further concerns. I've checked embedded WebView and MediaPlayer in Swing and a standalone WebView. Looks good.

**Comment by kevinrushforth** [comment]
Thanks for the reply. I'll go ahead and integrate it then.

**Pull request closed without merging**
